### PR TITLE
Add Dynamic Recurrent Basket Model (DREAM)

### PR DESCRIPTION
**Pull request by lthoang**
### Related Issues

https://github.com/PreferredAI/cornac/issues/579

### Checklist:

- [ ] I have added tests.
- [x] I have updated the documentation accordingly.
- [x] I have updated `README.md` (if you are adding a new model).
- [x] I have updated `examples/README.md` (if you are adding a new example).
- [ ] I have updated `datasets/README.md` (if you are adding a new dataset).

### CODE REVIEW

File: cornac/models/dream/dream.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+import torch
+import torch.nn as nn
+
+from torch.autograd import Variable
+from torch.nn.init import constant_, xavier_normal_
+from torch.nn.utils.clip_grad import clip_grad_norm_
+from tqdm.auto import trange
+
+
+class Wloss(nn.modules.loss._Loss):
```
Comment (tqtg) [inline cornac/models/dream/dream.py:14]: What is W loss? Can it be more informative or is it an acronym by the authors?

File: README.md [hunk 2]
```diff
@@ -177,6 +177,7 @@ The recommender models supported by Cornac are listed below. Why don't you join
 | 2016 | [Collaborative Deep Ranking (CDR)](cornac/models/cdr), [paper](http://inpluslab.com/chenliang/homepagefiles/paper/hao-pakdd2016.pdf) | [requirements.txt](cornac/models/cdr/requirements.txt) | [cdr_exp.py](examples/cdr_example.py)
 |      | [Collaborative Ordinal Embedding (COE)](cornac/models/coe), [paper](http://www.hadylauw.com/publications/sdm16.pdf) | [requirements.txt](cornac/models/coe/requirements.txt) |
 |      | [Convolutional Matrix Factorization (ConvMF)](cornac/models/conv_mf), [paper](http://uclab.khu.ac.kr/resources/publication/C_351.pdf) | [requirements.txt](cornac/models/conv_mf/requirements.txt) | [convmf_exp.py](examples/conv_mf_example.py)
+|      | [Dynamic REcurrent bAsket Model (DREAM)](cornac/models/dream), [paper](https://cseweb.ucsd.edu/classes/fa17/cse291-b/reading/A%20Dynamic%20Recurrent%20Model%20for%20Next%20Basket%20Recommendation.pdf) | [requirements.txt](cornac/models/dream/requirements.txt) | [dream_tafeng.py](examples/dream_tafeng.py)
```
Comment (tqtg) [inline README.md:180]: Let's just do first cap for the title so everything looks uniform

Comment (lthoang) [reply README.md]: Updated.

File: cornac/models/dream/dream.py (hunk 1 above)
Comment (lthoang) [reply cornac/models/dream/dream.py:14]: @tqtg I am not quite sure about the acronym for `Wloss`. The current implementation is based on this implemenation https://github.com/liming-7/A-Next-Basket-Recommendation-Reality-Check/tree/main/methods/dream

The authors of the above implementation also refer the running source code to this repo https://github.com/yihong-chen/DREAM

However, these two implementations are quite different in their naming convention. I haven't read through them thoroughly. Let's take some time to check the validity of both implementations, whether they reflect the [paper](https://cseweb.ucsd.edu/classes/fa17/cse291-b/reading/A%20Dynamic%20Recurrent%20Model%20for%20Next%20Basket%20Recommendation.pdf) idea.

Comment (tqtg) [reply cornac/models/dream/dream.py:14]: Glancing at the paper, I only see BPR loss is used without additional information. I think it's better to rely on the [original implementation](https://github.com/yihong-chen/DREAM/blob/master/src/train.py) which contains 2 variants (BPR and reordered BPR).

File: cornac/models/dream/dream.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+            num_item=self.n_items,
+            dropout_prob=self.dropout_prob,
+            attention=self.attention,
+            device=self.device,
+        )
+
+        self.loss_fct = nn.BCELoss()
+        self.p_loss_fct = Wloss(self.loss_uplift, 1)
+        self.n_loss_fct = Wloss(1, self.loss_uplift)
+        self.meta_loss_fct = nn.MSELoss()
```
Comment (tqtg) [inline cornac/models/dream/dream.py:105]: I think `meta_loss_fct` is not used anywhere during training.

**Comment by tqtg** [comment]
@lthoang are we actively working on this?

**Comment by lthoang** [comment]
> @lthoang are we actively working on this?

Yes. I will be back to this model next week.
